In [1]:
%load_ext sql

1. The World Bank's international debt data
It's not that we humans only take debts to manage our necessities. A country may also take debt to manage its economy. For example, infrastructure spending is one costly ingredient required for a country's citizens to lead comfortable lives. The World Bank is the organization that provides debt to countries.

In this notebook, we are going to analyze international debt data collected by The World Bank. The dataset contains information about the amount of debt (in USD) owed by developing countries across several categories. We are going to find the answers to questions like:

What is the total amount of debt that is owed by the countries listed in the dataset?
Which country owns the maximum amount of debt and what does that amount look like?
What is the average amount of debt owed by countries across different debt indicators?


The first line of code connects us to the international_debt database where the table international_debt is residing. Let's first SELECT all of the columns from the international_debt table. Also, we'll limit the output to the first ten rows to keep the output clean.

In [2]:
%load_ext sql
%sql postgresql://postgres:password@localhost:5432/postgres

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [3]:
%%sql
SELECT * FROM international_debt LIMIT 10;

 * postgresql://postgres:***@localhost:5432/postgres
10 rows affected.


[('Afghanistan', 'AFG', 'Disbursements on external debt, long-term (DIS, current US$)', 'DT.DIS.DLXF.CD', Decimal('72894453.700000003')),
 ('Afghanistan', 'AFG', 'Interest payments on external debt, long-term (INT, current US$)', 'DT.INT.DLXF.CD', Decimal('53239440.100000001')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (AMT, current US$)', 'DT.AMT.BLAT.CD', Decimal('61739336.899999999')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (DIS, current US$)', 'DT.DIS.BLAT.CD', Decimal('49114729.399999999')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (INT, current US$)', 'DT.INT.BLAT.CD', Decimal('39903620.100000001')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (AMT, current US$)', 'DT.AMT.MLAT.CD', Decimal('39107845')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (DIS, current US$)', 'DT.DIS.MLAT.CD', Decimal('23779724.300000001')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (INT, current US$)', 'DT.INT.MLAT.CD', Decimal('13335820')),
 ('Afghanistan', 'AFG', 'PPG, official creditors (AMT, current US$)', 'DT.AMT.OFFT.CD', Decimal('100847181.900000006')),
 ('Afghanistan', 'AFG', 'PPG, official creditors (DIS, current US$)', 'DT.DIS.OFFT.CD', Decimal('72894453.700000003'))]

2. Finding the number of distinct countries
From the first ten rows, we can see the amount of debt owed by Afghanistan in the different debt indicators. But we do not know the number of different countries we have on the table. There are repetitions in the country names because a country is most likely to have debt in more than one debt indicator.

Without a count of unique countries, we will not be able to perform our statistical analyses holistically. In this section, we are going to extract the number of unique countries present in the table.

In [4]:
%%sql
SELECT COUNT(DISTINCT(country_name)) AS Total_distinct_countries FROM international_debt;

 * postgresql://postgres:***@localhost:5432/postgres
1 rows affected.


[(124,)]

3. Finding out the distinct debt indicators
We can see there are a total of 124 countries present on the table. As we saw in the first section, there is a column called indicator_name that briefly specifies the purpose of taking the debt. Just beside that column, there is another column called indicator_code which symbolizes the category of these debts. Knowing about these various debt indicators will help us to understand the areas in which a country can possibly be indebted to.

In [5]:
%%sql
SELECT DISTINCT(indicator_code) AS distinct_debt_indicators FROM international_debt ORDER BY distinct_debt_indicators;

 * postgresql://postgres:***@localhost:5432/postgres
25 rows affected.


[('DT.AMT.BLAT.CD',),
 ('DT.AMT.DLXF.CD',),
 ('DT.AMT.DPNG.CD',),
 ('DT.AMT.MLAT.CD',),
 ('DT.AMT.OFFT.CD',),
 ('DT.AMT.PBND.CD',),
 ('DT.AMT.PCBK.CD',),
 ('DT.AMT.PROP.CD',),
 ('DT.AMT.PRVT.CD',),
 ('DT.DIS.BLAT.CD',),
 ('DT.DIS.DLXF.CD',),
 ('DT.DIS.MLAT.CD',),
 ('DT.DIS.OFFT.CD',),
 ('DT.DIS.PCBK.CD',),
 ('DT.DIS.PROP.CD',),
 ('DT.DIS.PRVT.CD',),
 ('DT.INT.BLAT.CD',),
 ('DT.INT.DLXF.CD',),
 ('DT.INT.DPNG.CD',),
 ('DT.INT.MLAT.CD',),
 ('DT.INT.OFFT.CD',),
 ('DT.INT.PBND.CD',),
 ('DT.INT.PCBK.CD',),
 ('DT.INT.PROP.CD',),
 ('DT.INT.PRVT.CD',)]

4. Totaling the amount of debt owed by the countries
As mentioned earlier, the financial debt of a particular country represents its economic state. But if we were to project this on an overall global scale, how will we approach it?

Let's switch gears from the debt indicators now and find out the total amount of debt (in USD) that is owed by the different countries. This will give us a sense of how the overall economy of the entire world is holding up.

In [6]:
%%sql
SELECT ROUND(SUM(debt)/1000000,2) AS "total_debt_USD(Millions)" FROM international_debt;

 * postgresql://postgres:***@localhost:5432/postgres
1 rows affected.


[(Decimal('3079734.49'),)]

5. Country with the highest debt
"Human beings cannot comprehend very large or very small numbers. It would be useful for us to acknowledge that fact." - Daniel Kahneman. That is more than 3 million million USD, an amount which is really hard for us to fathom.

Now that we have the exact total of the amounts of debt owed by several countries, let's now find out the country that owns the highest amount of debt along with the amount. Note that this debt is the sum of different debts owed by a country across several categories. This will help to understand more about the country in terms of its socio-economic scenarios. We can also find out the category in which the country owns its highest debt. But we will leave that for now.

In [7]:
%%sql
SELECT country_name, ROUND(SUM(debt)/1000000,2) AS "total_debt_USD(Millions)" FROM international_debt GROUP BY country_name ORDER BY "total_debt_USD(Millions)" DESC LIMIT 1;

 * postgresql://postgres:***@localhost:5432/postgres
1 rows affected.


[('China', Decimal('285793.49'))]

6. Average amount of debt across indicators
So, it was China. A more in-depth breakdown of China's debts can be found here.

We now have a brief overview of the dataset and a few of its summary statistics. We already have an idea of the different debt indicators in which the countries owe their debts. We can dig even further to find out on an average how much debt a country owes? This will give us a better sense of the distribution of the amount of debt across different indicators.

In [8]:
%%sql
SELECT indicator_code AS debt_indicator, ROUND(AVG(debt)/1000000,2) as "average_debt_usd(millions)", indicator_name FROM international_debt GROUP BY debt_indicator, indicator_name ORDER BY "average_debt_usd(millions)" DESC LIMIT 10;

 * postgresql://postgres:***@localhost:5432/postgres
10 rows affected.


[('DT.AMT.DLXF.CD', Decimal('5904.87'), 'Principal repayments on external debt, long-term (AMT, current US$)'),
 ('DT.AMT.DPNG.CD', Decimal('5161.19'), 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)'),
 ('DT.DIS.DLXF.CD', Decimal('2152.04'), 'Disbursements on external debt, long-term (DIS, current US$)'),
 ('DT.DIS.OFFT.CD', Decimal('1958.98'), 'PPG, official creditors (DIS, current US$)'),
 ('DT.AMT.PRVT.CD', Decimal('1803.69'), 'PPG, private creditors (AMT, current US$)'),
 ('DT.INT.DLXF.CD', Decimal('1644.02'), 'Interest payments on external debt, long-term (INT, current US$)'),
 ('DT.DIS.BLAT.CD', Decimal('1223.14'), 'PPG, bilateral (DIS, current US$)'),
 ('DT.INT.DPNG.CD', Decimal('1220.41'), 'Interest payments on external debt, private nonguaranteed (PNG) (INT, current US$)'),
 ('DT.AMT.OFFT.CD', Decimal('1191.19'), 'PPG, official creditors (AMT, current US$)'),
 ('DT.AMT.PBND.CD', Decimal('1082.62'), 'PPG, bonds (AMT, current US$)')]

7. The highest amount of principal repayments
We can see that the indicator DT.AMT.DLXF.CD tops the chart of average debt. This category includes repayment of long term debts. Countries take on long-term debt to acquire immediate capital. More information about this category can be found here.

An interesting observation in the above finding is that there is a huge difference in the amounts of the indicators after the second one. This indicates that the first two indicators might be the most severe categories in which the countries owe their debts.

We can investigate this a bit more so as to find out which country owes the highest amount of debt in the category of long term debts (DT.AMT.DLXF.CD). Since not all the countries suffer from the same kind of economic disturbances, this finding will allow us to understand that particular country's economic condition a bit more specifically.

In [9]:
%%sql
SELECT country_name, ROUND(AVG(debt)/1000000,2) AS "average_debt_usd(millions)", indicator_name, indicator_code FROM international_debt
GROUP BY country_name, indicator_name, indicator_code HAVING indicaTor_code = 'DT.AMT.DLXF.CD' 
ORDER BY "average_debt_usd(millions)" DESC LIMIT 10;

 * postgresql://postgres:***@localhost:5432/postgres
10 rows affected.


[('China', Decimal('96218.62'), 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Brazil', Decimal('90041.84'), 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Russian Federation', Decimal('66589.76'), 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Turkey', Decimal('51555.03'), 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('South Asia', Decimal('48756.30'), 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('India', Decimal('31923.51'), 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Indonesia', Decimal('30916.11'), 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Kazakhstan', Decimal('27482.09'), 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Mexico', Decimal('25218.50'), 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Least developed countries: UN classification', Decimal('25197.03'), 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD')]

8. The most common debt indicator
China has the highest amount of debt in the long-term debt (DT.AMT.DLXF.CD) category. This is verified by The World Bank. It is often a good idea to verify our analyses like this since it validates that our investigations are correct.

We saw that long-term debt is the topmost category when it comes to the average amount of debt. But is it the most common indicator in which the countries owe their debt? Let's find that out.

In [10]:
%%sql
SELECT indicator_name, COUNT(indicator_code) AS indicator_count FROM international_debt
GROUP BY indicator_name
ORDER BY indicator_count DESC
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/postgres
20 rows affected.


[('PPG, multilateral (INT, current US$)', 124),
 ('Principal repayments on external debt, long-term (AMT, current US$)', 124),
 ('Interest payments on external debt, long-term (INT, current US$)', 124),
 ('PPG, official creditors (INT, current US$)', 124),
 ('PPG, multilateral (AMT, current US$)', 124),
 ('PPG, official creditors (AMT, current US$)', 124),
 ('Disbursements on external debt, long-term (DIS, current US$)', 123),
 ('PPG, official creditors (DIS, current US$)', 122),
 ('PPG, bilateral (AMT, current US$)', 122),
 ('PPG, bilateral (INT, current US$)', 122),
 ('PPG, multilateral (DIS, current US$)', 120),
 ('PPG, bilateral (DIS, current US$)', 113),
 ('PPG, private creditors (AMT, current US$)', 98),
 ('PPG, private creditors (INT, current US$)', 98),
 ('PPG, commercial banks (INT, current US$)', 84),
 ('PPG, commercial banks (AMT, current US$)', 84),
 ('Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', 79),
 ('Interest payments on external debt, private nonguaranteed (PNG) (INT, current US$)', 79),
 ('PPG, bonds (AMT, current US$)', 69),
 ('PPG, bonds (INT, current US$)', 69)]

9. Other viable debt issues and conclusion
There are a total of six debt indicators in which all the countries listed in our dataset have taken debt. The indicator DT.AMT.DLXF.CD is also there in the list. So, this gives us a clue that all these countries are suffering from a common economic issue. But that is not the end of the story, a part of the story rather.

Let's change tracks from debt_indicators now and focus on the amount of debt again. Let's find out the maximum amount of debt across the indicators along with the respective country names. With this, we will be in a position to identify the other plausible economic issues a country might be going through. By the end of this section, we will have found out the debt indicators in which a country owes its highest debt.

In this notebook, we took a look at debt owed by countries across the globe. We extracted a few summary statistics from the data and unraveled some interesting facts and figures. We also validated our findings to make sure the investigations are correct.

In [11]:
%%sql
SELECT country_name, indicator_name, MAX(debt) AS maximum_debt FROM international_debt
GROUP BY country_name, indicator_name
ORDER BY maximum_debt DESC
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/postgres
20 rows affected.


[('China', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('96218620835.699996948')),
 ('Brazil', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('90041840304.100006104')),
 ('China', 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', Decimal('72392986213.800003052')),
 ('Russian Federation', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('66589761833.5')),
 ('Turkey', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('51555031005.800003052')),
 ('South Asia', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('48756295898.199996948')),
 ('Brazil', 'PPG, private creditors (AMT, current US$)', Decimal('43598697498.599998474')),
 ('Russian Federation', 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', Decimal('42800154974.900001526')),
 ('Brazil', 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', Decimal('41831444053.300003052')),
 ('Least developed countries: UN classification', 'Disbursements on external debt, long-term (DIS, current US$)', Decimal('40160766261.599998474')),
 ('Turkey', 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', Decimal('39916197430')),
 ('Least developed countries: UN classification', 'PPG, official creditors (DIS, current US$)', Decimal('37273977350.5')),
 ('Brazil', 'PPG, other private creditors (AMT, current US$)', Decimal('35119003750')),
 ('IDA only', 'Disbursements on external debt, long-term (DIS, current US$)', Decimal('34531188113.199996948')),
 ('IDA only', 'PPG, official creditors (DIS, current US$)', Decimal('33258230166.900001526')),
 ('India', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('31923507000.799999237')),
 ('Indonesia', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('30916112653.799999237')),
 ('South Asia', 'Disbursements on external debt, long-term (DIS, current US$)', Decimal('29306216064.700000763')),
 ('South Asia', 'PPG, official creditors (DIS, current US$)', Decimal('28991897351.5')),
 ('Kazakhstan', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('27482093686.400001526'))]

In [12]:
%sql postgresql://postgres:password@localhost:5432/postgres

In [13]:
%%sql
SELECT * FROM international_debt LIMIT 10;

 * postgresql://postgres:***@localhost:5432/postgres
10 rows affected.


[('Afghanistan', 'AFG', 'Disbursements on external debt, long-term (DIS, current US$)', 'DT.DIS.DLXF.CD', Decimal('72894453.700000003')),
 ('Afghanistan', 'AFG', 'Interest payments on external debt, long-term (INT, current US$)', 'DT.INT.DLXF.CD', Decimal('53239440.100000001')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (AMT, current US$)', 'DT.AMT.BLAT.CD', Decimal('61739336.899999999')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (DIS, current US$)', 'DT.DIS.BLAT.CD', Decimal('49114729.399999999')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (INT, current US$)', 'DT.INT.BLAT.CD', Decimal('39903620.100000001')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (AMT, current US$)', 'DT.AMT.MLAT.CD', Decimal('39107845')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (DIS, current US$)', 'DT.DIS.MLAT.CD', Decimal('23779724.300000001')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (INT, current US$)', 'DT.INT.MLAT.CD', Decimal('13335820')),
 ('Afghanistan', 'AFG', 'PPG, official creditors (AMT, current US$)', 'DT.AMT.OFFT.CD', Decimal('100847181.900000006')),
 ('Afghanistan', 'AFG', 'PPG, official creditors (DIS, current US$)', 'DT.DIS.OFFT.CD', Decimal('72894453.700000003'))]